File: src/ts/conformance.test.ts
// Runs the shared conformance corpus against the TypeScript implementation.
// The Rust and Dart slices run the same file; a case that passes in one
// language and fails in another is the drift this repository exists to catch.
//
//   node --test conformance.test.ts
//
// No build step: Node strips the types on the way in.

import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { test } from "node:test";

import type { PackageMetadata, VersionScheme } from "@zed-pkg/zed-interfaces";

import { ResolveError, resolveVersion } from "./index.ts";

interface Case {
  readonly name: string;
  readonly scheme: string;
  readonly versions: readonly string[];
  readonly requirement: string;
  readonly expect: { readonly version?: string; readonly error?: string };
}

const corpusPath = path.join(
  import.meta.dirname,
  "../../conformance/cases/version-resolution.json",
);
const corpus = JSON.parse(fs.readFileSync(corpusPath, "utf8")) as { cases: Case[] };

const metadataFor = (scheme: string, versions: readonly string[]): PackageMetadata => ({
  org: "acme",
  name: "conformance",
  vcs: "git",
  repo_url: "https://github.com/acme/conformance",
  latest: versions.length ? (versions[versions.length - 1] as string) : null,
  versions,
  version_scheme: (scheme === "calver" || scheme === "opaque" ? scheme : "semver") as VersionScheme,
});

test("the corpus is not empty", () => {
  assert.ok(corpus.cases.length > 0);
});

for (const testCase of corpus.cases) {
  test(testCase.name, () => {
    const metadata = metadataFor(testCase.scheme, testCase.versions);
    const { version: want, error: wantError } = testCase.expect;
    assert.ok(
      (want === undefined) !== (wantError === undefined),
      "a case declares exactly one of `version` or `error`",
    );

    if (want !== undefined) {
      assert.equal(resolveVersion(metadata, testCase.requirement), want);
      return;
    }
    assert.throws(
      () => resolveVersion(metadata, testCase.requirement),
      (error: unknown) => {
        assert.ok(error instanceof ResolveError);
        assert.equal(error.kind, wantError);
        return true;
      },
    );
  });
}

File: src/dart/test/conformance_test.dart
// Runs the shared conformance corpus against the Dart implementation. The Rust
// slice runs the same file; a case that passes in one language and fails in the
// other is the drift this repository exists to catch.

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:zed_interfaces/package_metadata.dart';
import 'package:zed_lib/zed_lib.dart';

const _corpusPath = '../../conformance/cases/version-resolution.json';

PackageMetadata metadataFor(String scheme, List<String> versions) =>
    PackageMetadata(
      org: 'acme',
      name: 'conformance',
      vcs: Vcs.git,
      repoUrl: 'https://github.com/acme/conformance',
      latest: versions.isEmpty ? null : versions.last,
      versions: versions,
      versionScheme: switch (scheme) {
        'calver' => VersionScheme.calver,
        'opaque' => VersionScheme.opaque,
        _ => VersionScheme.semver,
      },
    );

void main() {
  final raw = File(_corpusPath).readAsStringSync();
  final corpus = jsonDecode(raw) as Map<String, dynamic>;
  final cases = (corpus['cases'] as List<dynamic>).cast<Map<String, dynamic>>();

  test('the corpus is not empty', () => expect(cases, isNotEmpty));

  for (final testCase in cases) {
    final name = testCase['name'] as String;
    final metadata = metadataFor(
      testCase['scheme'] as String,
      (testCase['versions'] as List<dynamic>).cast<String>(),
    );
    final requirement = testCase['requirement'] as String;
    final expected = testCase['expect'] as Map<String, dynamic>;

    test(name, () {
      final wantVersion = expected['version'] as String?;
      final wantError = expected['error'] as String?;
      expect(
        (wantVersion == null) != (wantError == null),
        isTrue,
        reason: 'a case declares exactly one of `version` or `error`',
      );

      if (wantVersion != null) {
        expect(resolveVersion(metadata, requirement), wantVersion);
      } else {
        try {
          final resolved = resolveVersion(metadata, requirement);
          fail('expected $wantError, resolved $resolved');
        } on ResolveException catch (error) {
          expect(error.kind.wire, wantError);
        }
      }
    });
  }
}
